scipy/sparse/linalg/_interface.py

```diff
@@ -384,6 +384,8 @@ def rmatmat(self, X):
             raise ValueError(f'dimension mismatch: {self.shape}, {X.shape}')
 
         try:
+            if self._rmatmat is None:
+                raise TypeError("The rmatmat method is not defined for this LinearOperator.")
```

Comment:
I don't see a test for this in this PR yet; something like this, possibly adjusted to `NotImplementedError` as another reviewer notes, might make sense:

<details>

```diff
--- a/scipy/sparse/linalg/tests/test_interface.py
+++ b/scipy/sparse/linalg/tests/test_interface.py
@@ -4,6 +4,7 @@
 from functools import partial
 from itertools import product
 import operator
+import pytest
 from pytest import raises as assert_raises, warns
 from numpy.testing import assert_, assert_equal
 
@@ -11,6 +12,7 @@ import numpy as np
 import scipy.sparse as sparse
 
 import scipy.sparse.linalg._interface as interface
+from scipy.sparse.linalg import LinearOperator
 from scipy.sparse._sputils import matrix
 
 
@@ -479,3 +481,8 @@ def test_sparse_matmat_exception():
         A @ np.identity(4)
     with assert_raises(ValueError):
         np.identity(4) @ A
+
+
+def test_gh_18140():
+    with pytest.raises(TypeError, match="rmatmat method is not defined"):
+        LinearOperator(matvec=lambda x: x, shape=(1, 1)).rmatmat([[1]])
```

</details>

It sounds like Jake knows more than I about it, but I did notice that when I test your branch with my draft regression test above, not only does the error message mismatch to what you wrote, but I still get the original traceback that the matching issue complains about.